### ISSUE TITLE
timeout option for controls

### ISSUE THREAD

**Issue by rmoriz** [inspec/inspec#1675]
I have this nasty case when an command just hangs in the error case after outputting an error message. This also blocks the entire inspec run.

Example on Ubuntu 16.04:

- install package `duplicity`, then run inspect:

```ruby 
# tahoe
describe command('duplicity --dry-run /tmp tahoe://alias/directory') do
  its('stderr') { should_not match /pkg_resources.DistributionNotFound/ }
end
```
So I wish I had the following options:

- a general, inspec-global timeout for all resources
- a way to specify a timeout per resource (e.g. command)

Maybe something like:

```ruby 
# tahoe
describe command('duplicity --dry-run /tmp tahoe://alias/directory', timeout: 3) do
  its('stderr') { should_not match /pkg_resources.DistributionNotFound/ }
end
```

Currently i'm falling back to coreutil's `timeout`

```ruby 
# tahoe
describe command('timeout 3 duplicity --dry-run /tmp tahoe://alias/directory') do
  its('stderr') { should_not match /pkg_resources.DistributionNotFound/ }
end
```

**Comment by chris-rock** [comment]
I love that idea @rmoriz Each `describe` or `control` should have a default timeout that users can tweak if required. Instead of adding this feature to each resource, i'd like to have it as a general concept. This would allow us to abort any test if a defined timeout has reached. An option for controls could be:

```
control 'install' do
  describe command('duplicity --dry-run /tmp tahoe://alias/directory') do
    its('stderr') { should_not match /pkg_resources.DistributionNotFound/ }
   end
  timeout 3000 # in milliseconds
end
```
I need to think about a way to add this to `describe` blocks without `controls`. @rmoriz you are mainly doing integration tests instead of compliance tests right?

**Comment by chris-rock** [comment]
Once #1205 is in, we could raise a proper timeout exception

**Comment by adamleff** [comment]
I wonder if it would make more sense for the first pass of this to be a CLI flag to `inspec exec` that allows the user to provide a timeout to any `command` call and we then pass to train and enforce it there. It could then be exposed in `kitchen-inspec` as a driver configuration option too. It feels like that would address the need without a lot of complexity, and then we can further refine it with control-specific overrides if needed.

**Comment by chris-rock** [comment]
Sounds like a good question for @arlimus

**Comment by clintoncwolfe** [comment]
The OP's specific problem is being addressed on #3772 and #3771.  This issue is now about exploring how possible it would be to add a few different timeout features:

 * A per-resource timeout, as a resource parameter (like #3772, but for everything)
 * A per-control timeout, likely as a control DSL
 * A global timeout - perhaps a CLI option that restricts the entire run to finish in a certain time, or fail all remaining resources.

**Comment by clintoncwolfe** [comment]
Just to keep the discussion informed for the constraints, doing timeouts in Ruby is unusually difficult; see the [caveats](https://medium.com/@adamhooper/in-ruby-dont-use-timeout-77d9d4e5a001). The standard library is unsafe; and making timeout changes to each resource would require changes that are specific to the underlying libraries they use (which may be a win, if the libraries support timeouts, or a major obstacle if they don't). 

That said, this [library is a goldmine](https://github.com/ankane/the-ultimate-guide-to-ruby-timeouts) for implementation information.

**Comment by b-dean** [comment]
Also, it might be nice to have some way to set timeouts for connecting to the Train transport. I'm using an ssh target and the target is not necessarily up before Train's [default connection timeout + retries](https://github.com/inspec/train/blob/c93845e722998b6ea3c617064cd08d14f3f85f2a/lib/train/transports/ssh.rb#L55-L57) (should be 80 seconds for all the retires) expires. 

The `inspec exec` CLI doesn't give any options for configuring the transport options.

**Comment by clintoncwolfe** [comment]
Indeed, https://github.com/inspec/train/issues/352 tracks adding Train connection timeouts.

**Comment by kekaichinose** [comment]
Per @clintoncwolfe's comment on 1/30/2019, the OP's issue should be closed out. If there's a new issue that's been spawned as a follow-on to this one, please explain the details, including the benefit and value, in a new issue. Even better, we would love for you to participate by contributing a PR for the follow-on for us to review.

Thanks!
Chef InSpec Team

**Comment by galenemery** [comment]
In light of recent issues with controls taking a long time to complete, I'd like to revisit this issue as a broader implementation/discussion of how to handle a control that is taking too long.

The issue is not limited solely to the command resource.  We recently experienced this with WMI queries run through powershell.

@kekaichinose

**Comment by sean-horn** [comment]
Comes into play specifically in other cases too, like xccdf_org.cisecurity.benchmarks_rule_4.1.13_Ensure_use_of_privileged_commands_is_collected in the CIS profile when the filesystem being searched has hung NFS mounts.
